# Load Libraries and dataset

### Libraries

In [1]:
import pandas as pd
import numpy as np

# model building libraries
from sklearn.model_selection import train_test_split

from sklearn.metrics import mean_absolute_error

import statsmodels.api as sm
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.neural_network import MLPClassifier
from sklearn.metrics import precision_recall_fscore_support as score

import matplotlib.pyplot as plt


/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


### Prepare Data and create dummy variables

In [3]:
train = pd.read_csv('/content/drive/My Drive/DS_Projects/claims_severity/train_clean.csv')
#test = pd.read_csv('/content/drive/My Drive/DS_Projects/claims_severity/test.csv', error_bad_lines=False)

In [4]:
del_variables = list(train.columns)[119:122] + list(train.columns)[123:-1]
del_variables.append(train.columns[116])
del_variables.append(list(train.columns)[87])
print(del_variables)

['cont4', 'cont5', 'cont6', 'cont8', 'cont9', 'cont10', 'cont11', 'cont12', 'cont13', 'cont14', 'cont1', 'cat88']


In [5]:
dummy_variables = list(train.columns)[:87] + list(train.columns)[88:116]
print(dummy_variables)

['cat1', 'cat2', 'cat3', 'cat4', 'cat5', 'cat6', 'cat7', 'cat8', 'cat9', 'cat10', 'cat11', 'cat12', 'cat13', 'cat14', 'cat15', 'cat16', 'cat17', 'cat18', 'cat19', 'cat20', 'cat21', 'cat22', 'cat23', 'cat24', 'cat25', 'cat26', 'cat27', 'cat28', 'cat29', 'cat30', 'cat31', 'cat32', 'cat33', 'cat34', 'cat35', 'cat36', 'cat37', 'cat38', 'cat39', 'cat40', 'cat41', 'cat42', 'cat43', 'cat44', 'cat45', 'cat46', 'cat47', 'cat48', 'cat49', 'cat50', 'cat51', 'cat52', 'cat53', 'cat54', 'cat55', 'cat56', 'cat57', 'cat58', 'cat59', 'cat60', 'cat61', 'cat62', 'cat63', 'cat64', 'cat65', 'cat66', 'cat67', 'cat68', 'cat69', 'cat70', 'cat71', 'cat72', 'cat73', 'cat74', 'cat75', 'cat76', 'cat77', 'cat78', 'cat79', 'cat80', 'cat81', 'cat82', 'cat83', 'cat84', 'cat85', 'cat86', 'cat87', 'cat89', 'cat90', 'cat91', 'cat92', 'cat93', 'cat94', 'cat95', 'cat96', 'cat97', 'cat98', 'cat99', 'cat100', 'cat101', 'cat102', 'cat103', 'cat104', 'cat105', 'cat106', 'cat107', 'cat108', 'cat109', 'cat110', 'cat111', 'cat11

In [6]:
train = pd.get_dummies(train, columns=list(train.columns)[:87] + list(train.columns)[88:116])
train.drop(del_variables, axis=1, inplace=True)

In [7]:
#variables = dummy_variables + ['cont2', 'cont3', 'cont7', 'loss']

In [8]:
X_train, X_test, y_train, y_test = train_test_split(train.drop('loss', axis=1), train['loss'], test_size=0.2,
                                                   random_state=1234)

In [9]:
X_train.describe()

cont2          cont3  ...       cat116_X       cat116_Y
count  150654.000000  150654.000000  ...  150654.000000  150654.000000
mean        0.507241       0.498881  ...       0.000007       0.000060
std         0.207320       0.201752  ...       0.002576       0.007729
min         0.001149       0.002634  ...       0.000000       0.000000
25%         0.358319       0.336963  ...       0.000000       0.000000
50%         0.555782       0.527991  ...       0.000000       0.000000
75%         0.681761       0.634224  ...       0.000000       0.000000
max         0.862654       0.944251  ...       1.000000       1.000000

[8 rows x 1138 columns]

In [10]:
'cont2' in X_train.columns

True

In [11]:
X_train.shape

(150654, 1138)

In [12]:
y_train.shape

(150654,)

# Modeling

### Linear Model

In [ ]:
#X = sm.add_constant(X_train)
est = sm.OLS(y_train, X_train).fit()
est.summary()

/usr/local/lib/python3.6/dist-packages/statsmodels/base/model.py:1294: RuntimeWarning: invalid value encountered in true_divide
  return self.params / self.bse
/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:903: RuntimeWarning: invalid value encountered in greater
  return (a < x) & (x < b)
/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:903: RuntimeWarning: invalid value encountered in less
  return (a < x) & (x < b)
/usr/local/lib/python3.6/dist-packages/scipy/stats/_distn_infrastructure.py:1912: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= _a)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   loss   R-squared:                       0.531
Model:                            OLS   Adj. R-squared:                  0.528
Method:                 Least Squares   F-statistic:                     182.7
Date:                Sat, 19 Sep 2020   Prob (F-statistic):               0.00
Time:                        13:21:32   Log-Likelihood:            -1.3569e+06
No. Observations:              150654   AIC:                         2.716e+06
Df Residuals:                  149725   BIC:                         2.725e+06
Df Model:                         928                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
cont2       1353.1317     53.180     25.444      0.000    1248.900    1457.363
cont3        617.4173     99.297      6.218      0.000     422.797     812.038
cont7        684.0058    119.774      5.711      0.000     449.251     918.761
cat1_A       281.1237     18.717     15.020      0.000     244.439     317.808
cat1_B        42.6973     18.768      2.275      0.023       5.913      79.482
cat2_A       530.2458    191.671      2.766      0.006     154.575     905.917
cat2_B      -206.4248    191.405     -1.078      0.281    -581.574     168.724
cat3_A      -118.4105     96.946     -1.221      0.222    -308.423      71.602
cat3_B       442.2315    100.308      4.409      0.000     245.630     638.833
cat4_A       -49.5295    131.490     -0.377      0.706    -307.248     208.189
cat4_B       373.3504    131.110      2.848      0.004     116.378     630.323
cat5_A      -334.1886     80.763     -4.138      0.000    -492.482    -175.895
cat5_B       658.0096     79.609      8.266      0.000     501.978     814.041
cat6_A      -398.9256    533.613     -0.748      0.455   -1444.797     646.945
cat6_B       722.7466    518.399      1.394      0.163    -293.304    1738.797
cat7_A       -24.1559    120.318     -0.201      0.841    -259.977     211.665
cat7_B       347.9769    117.060      2.973      0.003     118.542     577.412
cat8_A      -168.5924    131.745     -1.280      0.201    -426.811      89.626
cat8_B       492.4134    130.436      3.775      0.000     236.762     748.065
cat9_A      -205.7326     78.982     -2.605      0.009    -360.536     -50.929
cat9_B       529.5536     81.603      6.489      0.000     369.613     689.494
cat10_A     -290.6070     78.346     -3.709      0.000    -444.164    -137.050
cat10_B      614.4280     80.977      7.588      0.000     455.715     773.142
cat11_A     -269.4900     78.296     -3.442      0.001    -422.949    -116.031
cat11_B      593.3110     80.945      7.330      0.000     434.660     751.962
cat12_A     -558.3687     78.367     -7.125      0.000    -711.966    -404.771
cat12_B      882.1897     80.980     10.894      0.000     723.470    1040.910
cat13_A     -265.3495     78.336     -3.387      0.001    -418.886    -111.813
cat13_B      589.1705     80.973      7.276      0.000     430.465     747.876
cat14_A     -360.1893     96.973     -3.714      0.000    -550.255    -170.124
cat14_B      684.0103    101.400      6.746      0.000     485.269     882.752
cat15_A      181.0566    199.950      0.906      0.365    -210.840     572.954
cat15_B      142.7644    204.616      0.698      0.485    -258.278     543.807
cat16_A      457.0212     88.427      5.168      0.000     283.706     630.336
cat16_B     -133.2002     89.742     -1.484      0.138    -309.093      42.693
cat17_A      346.3209     91.114      3.801      0.000     167.738     524.903
ca

Adjusted R-Square: 0.531

In [ ]:
y_pred = est.predict(X_test)

In [ ]:
mean_absolute_error(y_test, y_pred)

1304.9119840171602

In [ ]:
y_pred.mean()

3025.4123995341824

### Random Forest Regressor

In [ ]:
from sklearn.ensemble import RandomForestRegressor

In [ ]:
rg1 = RandomForestRegressor(n_jobs=-1, random_state=2525)

rg1.fit(X_train,y_train)

y_pred = rg1.predict(X_test)
mean_absolute_error(y_test, y_pred)

1245.3555110879372

In [ ]:
rg2 = RandomForestRegressor(n_estimators=120, n_jobs=-1, random_state=2525)

rg2.fit(X_train,y_train)

y_pred = rg2.predict(X_test)
mean_absolute_error(y_test, y_pred)

1244.3078535577977

### Gradient Boosting Regressor

In [14]:
from sklearn.ensemble import GradientBoostingRegressor

In [16]:
gb1 = GradientBoostingRegressor(random_state=2525)
gb1.fit(X_train,y_train)

GradientBoostingRegressor(alpha=0.9, ccp_alpha=0.0, criterion='friedman_mse',
                          init=None, learning_rate=0.1, loss='ls', max_depth=3,
                          max_features=None, max_leaf_nodes=None,
                          min_impurity_decrease=0.0, min_impurity_split=None,
                          min_samples_leaf=1, min_samples_split=2,
                          min_weight_fraction_leaf=0.0, n_estimators=100,
                          n_iter_no_change=None, presort='deprecated',
                          random_state=2525, subsample=1.0, tol=0.0001,
                          validation_fraction=0.1, verbose=0, warm_start=False)

In [17]:
y_pred = gb1.predict(X_test)
mean_absolute_error(y_test, y_pred)

1266.7901906644993

GB Model 2

In [18]:
gb2 = GradientBoostingRegressor(random_state=2525, n_estimators=200)
gb2.fit(X_train,y_train)

GradientBoostingRegressor(alpha=0.9, ccp_alpha=0.0, criterion='friedman_mse',
                          init=None, learning_rate=0.1, loss='ls', max_depth=3,
                          max_features=None, max_leaf_nodes=None,
                          min_impurity_decrease=0.0, min_impurity_split=None,
                          min_samples_leaf=1, min_samples_split=2,
                          min_weight_fraction_leaf=0.0, n_estimators=200,
                          n_iter_no_change=None, presort='deprecated',
                          random_state=2525, subsample=1.0, tol=0.0001,
                          validation_fraction=0.1, verbose=0, warm_start=False)

In [19]:
y_pred = gb2.predict(X_test)
mean_absolute_error(y_test, y_pred)

1237.6168790854867

GB: Model 3

In [20]:
gb3 = GradientBoostingRegressor(random_state=2525, n_estimators=300)
gb3.fit(X_train,y_train)

GradientBoostingRegressor(alpha=0.9, ccp_alpha=0.0, criterion='friedman_mse',
                          init=None, learning_rate=0.1, loss='ls', max_depth=3,
                          max_features=None, max_leaf_nodes=None,
                          min_impurity_decrease=0.0, min_impurity_split=None,
                          min_samples_leaf=1, min_samples_split=2,
                          min_weight_fraction_leaf=0.0, n_estimators=300,
                          n_iter_no_change=None, presort='deprecated',
                          random_state=2525, subsample=1.0, tol=0.0001,
                          validation_fraction=0.1, verbose=0, warm_start=False)

In [21]:
y_pred = gb3.predict(X_test)
mean_absolute_error(y_test, y_pred)

1230.1001412172816

GB: Model 4

In [22]:
gb4 = GradientBoostingRegressor(random_state=2525, n_estimators=400)
gb4.fit(X_train,y_train)

GradientBoostingRegressor(alpha=0.9, ccp_alpha=0.0, criterion='friedman_mse',
                          init=None, learning_rate=0.1, loss='ls', max_depth=3,
                          max_features=None, max_leaf_nodes=None,
                          min_impurity_decrease=0.0, min_impurity_split=None,
                          min_samples_leaf=1, min_samples_split=2,
                          min_weight_fraction_leaf=0.0, n_estimators=400,
                          n_iter_no_change=None, presort='deprecated',
                          random_state=2525, subsample=1.0, tol=0.0001,
                          validation_fraction=0.1, verbose=0, warm_start=False)

In [23]:
y_pred = gb4.predict(X_test)
mean_absolute_error(y_test, y_pred)

1226.1763225726934

GB: Model 5

In [24]:
gb5 = GradientBoostingRegressor(random_state=2525, n_estimators=500)
gb5.fit(X_train,y_train)

GradientBoostingRegressor(alpha=0.9, ccp_alpha=0.0, criterion='friedman_mse',
                          init=None, learning_rate=0.1, loss='ls', max_depth=3,
                          max_features=None, max_leaf_nodes=None,
                          min_impurity_decrease=0.0, min_impurity_split=None,
                          min_samples_leaf=1, min_samples_split=2,
                          min_weight_fraction_leaf=0.0, n_estimators=500,
                          n_iter_no_change=None, presort='deprecated',
                          random_state=2525, subsample=1.0, tol=0.0001,
                          validation_fraction=0.1, verbose=0, warm_start=False)

In [26]:
y_pred = gb5.predict(X_test)
mean_absolute_error(y_test, y_pred)

1223.5805546267857

GB: Model 6

In [27]:
gb6 = GradientBoostingRegressor(random_state=2525, n_estimators=500, learning_rate=0.05)
gb6.fit(X_train,y_train)

GradientBoostingRegressor(alpha=0.9, ccp_alpha=0.0, criterion='friedman_mse',
                          init=None, learning_rate=0.05, loss='ls', max_depth=3,
                          max_features=None, max_leaf_nodes=None,
                          min_impurity_decrease=0.0, min_impurity_split=None,
                          min_samples_leaf=1, min_samples_split=2,
                          min_weight_fraction_leaf=0.0, n_estimators=500,
                          n_iter_no_change=None, presort='deprecated',
                          random_state=2525, subsample=1.0, tol=0.0001,
                          validation_fraction=0.1, verbose=0, warm_start=False)

In [28]:
y_pred = gb6.predict(X_test)
mean_absolute_error(y_test, y_pred)

1231.8781595516214

In [33]:
!pip install scikit-optimize

     |████████████████████████████████| 102kB 2.3MB/s 


#### Hyper parameter estimation for Gradient Boosting

In [43]:
from skopt.space import Real, Integer
from skopt.utils import use_named_args

n_features = X_train.shape[1]
space  = [Integer(1, 5, name='max_depth'),
          Real(10**-5, 10**0, "log-uniform", name='learning_rate'),
          Integer(1, n_features, name='max_features'),
          Integer(2, 100, name='min_samples_split'),
          Integer(1, 100, name='min_samples_leaf')]

# this decorator allows your objective function to receive a the parameters as
# keyword arguments. This is particularly convenient when you want to set
# scikit-learn estimator parameters
@use_named_args(space)
def objective(**params):
    gb1.set_params(**params)

    return -np.mean(cross_val_score(gb1, X_train, y_train, cv=5, n_jobs=-1,
                                    scoring="neg_mean_absolute_error"))


In [44]:
from sklearn.model_selection import cross_val_score
from skopt import gp_minimize
res_gp = gp_minimize(objective, space, n_calls=50)

"Best score=%.4f" % res_gp.fun


'Best score=1214.5313'

In [45]:
print("""Best parameters:
- max_depth=%d
- learning_rate=%.6f
- max_features=%d
- min_samples_split=%d
- min_samples_leaf=%d""" % (res_gp.x[0], res_gp.x[1],
                            res_gp.x[2], res_gp.x[3],
                            res_gp.x[4]))

Best parameters:
- max_depth=5
- learning_rate=0.358839
- max_features=173
- min_samples_split=25
- min_samples_leaf=91


### Multi-layer Perceptron regressor (Neural Network)


In [13]:
from sklearn.neural_network import MLPRegressor

MLPR Model 1

In [ ]:
nn1 = MLPRegressor(random_state=1)
nn1.fit(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPRegressor(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
             beta_2=0.999, early_stopping=False, epsilon=1e-08,
             hidden_layer_sizes=(100,), learning_rate='constant',
             learning_rate_init=0.001, max_fun=15000, max_iter=200,
             momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
             power_t=0.5, random_state=1, shuffle=True, solver='adam',
             tol=0.0001, validation_fraction=0.1, verbose=False,
             warm_start=False)

In [ ]:
y_pred_nn1 = nn1.predict(X_test)
nn1_result = mean_absolute_error(y_test, y_pred_nn1)
print(nn1_result)

1202.3596050580013


MLPR Model 2

In [15]:
# add one more layer
nn2 =MLPRegressor(hidden_layer_sizes=(100,100,),random_state=1)
nn2.fit(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPRegressor(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
             beta_2=0.999, early_stopping=False, epsilon=1e-08,
             hidden_layer_sizes=(100, 100), learning_rate='constant',
             learning_rate_init=0.001, max_fun=15000, max_iter=200,
             momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
             power_t=0.5, random_state=1, shuffle=True, solver='adam',
             tol=0.0001, validation_fraction=0.1, verbose=False,
             warm_start=False)

In [16]:
y_pred_nn2 = nn2.predict(X_test)
nn2_result = mean_absolute_error(y_test, y_pred_nn2)
print(nn2_result)

1309.0007252635398


Actually by adding layer, the result got worsen. It may have overfit to the training dataset.

NN Model 3

In [15]:
nn3 =MLPRegressor(hidden_layer_sizes=(30,30,),random_state=1, alpha=0.001)
nn3.fit(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPRegressor(activation='relu', alpha=0.001, batch_size='auto', beta_1=0.9,
             beta_2=0.999, early_stopping=False, epsilon=1e-08,
             hidden_layer_sizes=(30, 30), learning_rate='constant',
             learning_rate_init=0.001, max_fun=15000, max_iter=200,
             momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
             power_t=0.5, random_state=1, shuffle=True, solver='adam',
             tol=0.0001, validation_fraction=0.1, verbose=False,
             warm_start=False)

In [19]:
y_pred_nn3 = nn3.predict(X_test)
nn3_result = mean_absolute_error(y_test, y_pred_nn3)
print(nn3_result)

1219.7420347587358


In [14]:
nn4 = MLPRegressor(random_state=1, hidden_layer_sizes=(80,))
nn4.fit(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPRegressor(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
             beta_2=0.999, early_stopping=False, epsilon=1e-08,
             hidden_layer_sizes=(80,), learning_rate='constant',
             learning_rate_init=0.001, max_fun=15000, max_iter=200,
             momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
             power_t=0.5, random_state=1, shuffle=True, solver='adam',
             tol=0.0001, validation_fraction=0.1, verbose=False,
             warm_start=False)

In [15]:
y_pred_nn4 = nn4.predict(X_test)
nn4_result = mean_absolute_error(y_test, y_pred_nn4)
print(nn4_result)

1193.0783006516408


Let's keep size down the hidden layers

In [16]:
nn5 = MLPRegressor(random_state=1, hidden_layer_sizes=(70,))
nn5.fit(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPRegressor(activation='relu', alpha=0.0001, batch_size='auto', beta_1=0.9,
             beta_2=0.999, early_stopping=False, epsilon=1e-08,
             hidden_layer_sizes=(70,), learning_rate='constant',
             learning_rate_init=0.001, max_fun=15000, max_iter=200,
             momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
             power_t=0.5, random_state=1, shuffle=True, solver='adam',
             tol=0.0001, validation_fraction=0.1, verbose=False,
             warm_start=False)

In [17]:
y_pred_nn5 = nn5.predict(X_test)
nn5_result = mean_absolute_error(y_test, y_pred_nn5)
print(nn5_result)

1193.1677359163489


In [19]:
nn6 = MLPRegressor(random_state=1, hidden_layer_sizes=(80,), activation='tanh')
nn6.fit(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPRegressor(activation='tanh', alpha=0.0001, batch_size='auto', beta_1=0.9,
             beta_2=0.999, early_stopping=False, epsilon=1e-08,
             hidden_layer_sizes=(80,), learning_rate='constant',
             learning_rate_init=0.001, max_fun=15000, max_iter=200,
             momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
             power_t=0.5, random_state=1, shuffle=True, solver='adam',
             tol=0.0001, validation_fraction=0.1, verbose=False,
             warm_start=False)

In [20]:
y_pred_nn6 = nn6.predict(X_test)
nn6_result = mean_absolute_error(y_test, y_pred_nn6)
print(nn6_result)

1241.6729772426604


In [21]:
nn7 = MLPRegressor(random_state=1, hidden_layer_sizes=(80,), alpha=0.001)
nn7.fit(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPRegressor(activation='relu', alpha=0.001, batch_size='auto', beta_1=0.9,
             beta_2=0.999, early_stopping=False, epsilon=1e-08,
             hidden_layer_sizes=(80,), learning_rate='constant',
             learning_rate_init=0.001, max_fun=15000, max_iter=200,
             momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
             power_t=0.5, random_state=1, shuffle=True, solver='adam',
             tol=0.0001, validation_fraction=0.1, verbose=False,
             warm_start=False)

In [22]:
y_pred_nn7 = nn7.predict(X_test)
nn7_result = mean_absolute_error(y_test, y_pred_nn7)
print(nn7_result)

1196.2222961903192


In [23]:
nn8 = MLPRegressor(random_state=1, hidden_layer_sizes=(80,), alpha=0.01)
nn8.fit(X_train, y_train)

/usr/local/lib/python3.6/dist-packages/sklearn/neural_network/_multilayer_perceptron.py:571: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (200) reached and the optimization hasn't converged yet.
  % self.max_iter, ConvergenceWarning)


MLPRegressor(activation='relu', alpha=0.01, batch_size='auto', beta_1=0.9,
             beta_2=0.999, early_stopping=False, epsilon=1e-08,
             hidden_layer_sizes=(80,), learning_rate='constant',
             learning_rate_init=0.001, max_fun=15000, max_iter=200,
             momentum=0.9, n_iter_no_change=10, nesterovs_momentum=True,
             power_t=0.5, random_state=1, shuffle=True, solver='adam',
             tol=0.0001, validation_fraction=0.1, verbose=False,
             warm_start=False)

In [24]:
y_pred_nn8 = nn8.predict(X_test)
nn8_result = mean_absolute_error(y_test, y_pred_nn8)
print(nn8_result)

1196.0794979140849


# Summary

We are evaluating models based on the mean absolute error rate for each model. <br>
Linear Regression Model - as a bench mark
- Score: 1304.912 

Random Forest Regressor
-  Best Score: 1244.31
-  It is better than linear regression

Gradient Boosting Regressor
- Best score: 1214.53
- Better than Linear Regression and Random Forest

Multi-layer Perception Regressor
- Best score: 1193.08
- With one hidden layer of 80 neurons